clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4994,10 +5003,14 @@ static void EmitSpecIdShims(raw_ostream &OS, unsigned &ShimCounter,
 // Returns a string containing the FQN of the 'top most' shim, including its
 // function call parameters.
 static std::string EmitSpecIdShims(raw_ostream &OS, unsigned &ShimCounter,
-                                   const VarDecl *VD) {
+                                   PrintingPolicy &Policy, const VarDecl *VD) {
   if (!VD->isInAnonymousNamespace())
     return "";
-  std::string RelativeName = VD->getNameAsString();
+  std::string RelativeName;
+  llvm::raw_string_ostream stream(RelativeName);
+  VD->getNameForDiagnostic(stream, Policy, false);
```

Comment:
Why do we want to use this API when we're not emitting a diagnostic? My fear is that this name may not be "real code" -- e.g., perhaps the diagnostic form elides some information that is unimportant for diagnostics but is important for source to source transformation, it may skip some extra adornments that are important (like use of the `typename` or `template` keywords), etc. I don't *know* that this happens now, but it could also be changed in the future by upstream.

I sort of wonder whether we should be using the pretty printer for this functionality instead? e.g., `VD->print()`.

Reply:
Because that is what I was able to find that actually printed the template arguments :)  If `VD->print` can accomplish this, I'm all for it.

Reply:
So after more exploration, it looks like this would be rather involved. The `DeclPrinter` has some printing policy options, but none to control printing of the type of the variable, or pragmas/attributes on it, etc. `printName()` was also an option, but that doesn't take a printing policy so we can't control what part of the name gets printed either.

So while my concern still exists, I think we should wait to do anything about it until we find a situation where it does the wrong thing.